packages/insomnia-app/app/ui/redux/selectors.ts

```
@@ -135,6 +132,30 @@ export const selectActiveWorkspaceMeta = createSelector(
   },
 );
 
+export const selectAllApiSpecs = createSelector(
+  selectEntitiesLists,
+  entities => entities.apiSpecs,
+);
+
+export const selectActiveApiSpec = createSelector(
+  selectAllApiSpecs,
+  selectActiveWorkspace,
+  (apiSpecs, activeWorkspace) => apiSpecs.find(apiSpec => apiSpec.parentId === activeWorkspace?._id) as ApiSpec
```

Please remove the cast and handle explicitly. Casting here is basically saying "ignore the error".

The same way it's possible for `selectActiveWorkspace` to return `undefined`, it's a totally valid state where there is no `activeApiSpec`.

Think about it: if there's no `activeWorkspace`, should there be an `activeApiSpec`? And, if there _is_ an `activeWorkspace`, we know as a business rule that there must be an `activeApiSpec`, so instead of casting we should throw an error knowing it's unreachable, and return `undefined` when we know it should be `undefined`.

Something like:

```ts
if (!activeWorkspace) {
    // There should never be an active api spec without an active workspace
    return undefined;
}

const activeSpec = apiSpecs.find(apiSpec => apiSpec.parentId === activeWorkspace._id);

if (!activeSpec) {
    // This case should never be reached; an api spec should always exist for a given workspace.
    throw new Error('xyz');
}

return activeSpec;
```

---

> Think about it: if there's no `activeWorkspace`, should there be an `activeApiSpec`? And, if there _is_ an `activeWorkspace`, we know as a business rule that there must be an `activeApiSpec`

🤔 afaik if the workspace scope is 'collection' it wouldn't have an ApiSpec associated with it?

---

In the current state the model is still always created, to avoid the discrepancy between design/collection from existing in the data layer and further complicating that problem. It's just that m in the UI, there is no way to see the apiSpec (no tab switcher at the top) in a request collection.

---

removing the cast I'm totally game for - and I did so in 60a87fc886d16bc6df37983085a2a1511ff9774c.

But this change to throw the error isn't really related to the scope of this ticket.  I've done what's strictly asked for up here in 495560ea4a659a07ee48120bc074783f10ab2920 but I'd ask that if we want to expand scope further that we do so in a separate ticket.  I know there's some quirky stuff going on with API specs - but we can't do it all at once and this ticket is about how a workspace name appears.  It's definitely unfortunate how the workspaces and apispecs are conflated and I'm glad we're taking steps towards fixing it -> but I'd like to cap it here for this PR if we can.

---

This particular example is not expanding scope; we introduced a new selector to achieve the desired behavior and that new selector logic doesn't handle all the logical codepaths.

Moreover, if that error state _is_ being hit, we should expect things to explode in our testing, because as mentioned in the comment that case should never be reached.

Thanks for adding it in, there's no more to be done for this selector.